core/src/Instance/B2CAuthority.cs

```diff
@@ -68,16 +68,28 @@ internal B2CAuthority(string authority, bool validateAuthority)
                 throw new ArgumentException(CoreErrorMessages.UnsupportedAuthorityValidation);
             }
 
-            var metadata = await AadInstanceDiscovery
+            InstanceDiscoveryMetadataEntry metadata = await GetB2CInstanceDiscoveryMetadataEntryAsync(
+                httpManager,
+                telemetryManager,
+                requestContext).ConfigureAwait(false);
+
+            CanonicalAuthority = UpdateHost(CanonicalAuthority, metadata.PreferredNetwork);
+        }
+
+        internal async Task<InstanceDiscoveryMetadataEntry> GetB2CInstanceDiscoveryMetadataEntryAsync(
+            IHttpManager httpManager,
+            ITelemetryManager telemetryManager,
+            RequestContext requestContext)
+        {
+             InstanceDiscoveryMetadataEntry metadata = await AadInstanceDiscovery
```

Comment:
Inheritance chains are difficult to maintain, because you have to keep track of what each base class is doing, if you are calling into the base class itself etc. I much prefer interfaces and encapsulation...

In this specific case, I see 2 problems: 

- InTrustedHostList is a private instance member in this class, and static in base class. You are actually calling into the base class, but this is difficult to comprehend. I would rename the static method as IsTrustedAADHost(), or move it to a separate class. 
- There is confusion in having a CannonicalAuthority field in the inheritance chain, which is updated A LOT during the lifetime of the class. And you add a GetCanonicalAuthorityB2C() which seems to be doing more changes. Can't all of these be consolidated ?

Reply:
outdated due to latest changes. (but i appreciate and did read/think about your comment :) )

---
In reply to: [234184038](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1405#discussion_r234184038) [](ancestors = 234184038)